linera-ethereum/src/client.rs

```diff
@@ -1,112 +1,304 @@
+                                return Err(serde::de::Error::duplicate_field("result"));
+                            }
+
+                            let value: &RawValue = map.next_value()?;
+                            result = Some(value);
+                        }
+                        "error" => {
+                            if error.is_some() {
+                                return Err(serde::de::Error::duplicate_field("error"));
+                            }
+
+                            let value: JsonRpcError = map.next_value()?;
+                            error = Some(value);
+                        }
+                        key => {
+                            return Err(serde::de::Error::unknown_field(
+                                key,
+                                &["id", "jsonrpc", "result", "error"],
+                            ))
+                        }
+                    }
+                }
+
+                // jsonrpc version must be present in all responses
+                if !jsonrpc {
+                    return Err(serde::de::Error::missing_field("jsonrpc"));
+                }
+
+                match (id, result, error) {
+                    (Some(id), Some(result), None) => Ok(Response::Success { id, result }),
+                    (Some(id), None, Some(error)) => Ok(Response::Error { id, error }),
+                    _ => Err(serde::de::Error::custom(
+                        "response must be either a success/error or notification object",
+                    )),
+                }
+            }
+        }
+
+        deserializer.deserialize_map(ResponseVisitor(&()))
+    }
+}
+
+/// The basic Ethereum queries that can be used from a smart contract and do not require
+/// gas to be executed.
+#[async_trait]
+pub trait EthereumQueries {
+    type Error;
+
+    /// Lists all the accounts of the Ethereum node.
+    async fn get_accounts(&self) -> Result<Vec<String>, Self::Error>;
+
+    /// Gets the latest block number of the Ethereum node.
+    async fn get_block_number(&self) -> Result<u64, Self::Error>;
 
     /// Gets the balance of the specified address at the specified block number.
     /// if no block number is specified then the balance of the latest block is
     /// returned.
-    pub async fn get_balance(
+    async fn get_balance(
         &self,
         address: &str,
         block_number: Option<u64>,
-    ) -> Result<U256, EthereumServiceError> {
-        let address = address.parse::<Address>()?;
-        let number = match block_number {
-            None => BlockNumberOrTag::Latest,
-            Some(val) => BlockNumberOrTag::Number(val),
-        };
-        let block_id = BlockId::Number(number);
-        Ok(self.provider.get_balance(address, block_id).await?)
-    }
+    ) -> Result<U256, Self::Error>;
 
     /// Reads the events of the smart contract.
     /// This is done from a specified `contract_address` and `event_name_expanded`.
```

Comment:
I think markdown only separates paragraphs if you insert a new line, e.g.:

```suggestion
    /// Reads the events of the smart contract.
    ///
    /// This is done from a specified `contract_address` and `event_name_expanded`.
```